Source/src/components/ComponentBillboard.cpp

```diff
@@ -0,0 +1,549 @@
+    {
+        time = 0.0f;
+        color_frame = 0.0f;
+    }
+}
+
+void Hachiko::ComponentBillboard::PublishIntoScene()
+{
+    auto scene = game_object->scene_owner;
+
+    if (scene)
+    {
+        scene->AddParticleComponent(this);
+        in_scene = true;
+    }
+}
+
+void Hachiko::ComponentBillboard::DetachFromScene()
+{
+    auto scene = game_object->scene_owner;
+
+    if (scene)
+    {
+        scene->RemoveParticleComponent(GetID());
+        in_scene = false;
+    }
+}
+
+void Hachiko::ComponentBillboard::GetOrientationMatrix(ComponentCamera* camera, float4x4& model_matrix)
+{
+    ComponentTransform* transform = GetGameObject()->GetComponent<ComponentTransform>();
+
+    Frustum* frustum = camera->GetFrustum();
+    float4x4* proj = &camera->GetProjectionMatrix();
+    float4x4* view = &camera->GetViewMatrix();
+
+    float3x3 rotatePart = transform->GetGlobalMatrix().RotatePart();
+    float3 position = transform->GetGlobalPosition();
+    float4x4 modelMatrix = transform->GetGlobalMatrix();
+    float3 scale = transform->GetGlobalScale();
+
+    if (type == BillboardType::NORMAL)
+    {
+        model_matrix = modelMatrix.LookAt(rotatePart.Col(2), -frustum->Front(), rotatePart.Col(1), float3::unitY);
+        model_matrix = float4x4::FromTRS(position, model_matrix.RotatePart() * rotatePart, scale);
+    }
+    else if (type == BillboardType::STRETCH)
+    {
+        float3 cameraPos = camera->GetFrustum()->Pos();
+        float3 cameraDir = (cameraPos - position).Normalized();
+        float3 upDir = Cross(direction, cameraDir);
+        float3 newCameraDir = Cross(direction, upDir);
+
+        float3x3 newRotation;
+        newRotation.SetCol(0, upDir);
+        newRotation.SetCol(1, direction);
+        newRotation.SetCol(2, newCameraDir);
+
+        model_matrix = float4x4::FromTRS(position, newRotation * model_stretch, scale);
+    }
+    else if (type == BillboardType::HORIZONTAL)
+    {
+        if (is_horizontal)
+        {
+            float3 direction = transform->GetGlobalRotation().WorldZ();
+            float3 projection = position + direction - direction.y * float3::unitY;
+            direction = (projection - position).Normalized();
+            float3 right = Cross(float3::unitY, direction);
+
+            float3x3 newRotation;
+            newRotation.SetCol(1, right);
+            newRotation.SetCol(2, float3::unitY);
+            newRotation.SetCol(0, direction);
+
+            model_matrix = float4x4::FromTRS(position, newRotation, scale);
+        }
+        else
+        {
+            model_matrix = float4x4::LookAt(float3::unitZ, float3::unitY, float3::unitY, float3::unitY);
+            model_matrix = float4x4::FromTRS(position, model_matrix.RotatePart(), scale);
+        }
+    }
+    else if (type == BillboardType::VERTICAL)
+    {
+        float3 cameraPos = camera->GetFrustum()->Pos();
+        float3 cameraDir = (float3(cameraPos.x, position.y, cameraPos.z) - position).Normalized();
```

Comment:
These two should be named like `camera_position` and `camera_direction`.

Reply:
Done